plutus-core/exe/Main.hs

```diff
@@ -412,15 +431,86 @@ runExample (ExampleOptions (ExampleSingle name)) = do
         Just ex -> PLC.render $ prettyExample ex
 
 
+---------------- Typechecking ----------------
+
+runTypecheck :: TypecheckOptions -> IO ()
+runTypecheck (TypecheckOptions inp fmt) = do
+  TypedProgram prog <- getProg Typed inp fmt
+  case PLC.runQuoteT $ do
+    types <- PLC.getStringBuiltinTypes ()
+    PLC.typecheckPipeline (PLC.TypeCheckConfig types) (void prog)
+    of
+       Left (e :: PLC.Error PLC.DefaultUni ()) -> T.putStrLn (PLC.displayPlcDef e) >> exitFailure
+       Right ty                                -> T.putStrLn (PLC.displayPlcDef ty) >> exitSuccess
+
+
+---------------- Evaluation ----------------
+
+runEval :: EvalOptions -> IO ()
+runEval (EvalOptions typing inp mode fmt printtime) =
+    case typing of
+      Typed -> do
+        TypedProgram prog <- getProg Typed inp fmt
+        let evaluate = case mode of
+                          CK  -> PLC.unsafeEvaluateCk  (PLC.getStringBuiltinMeanings @ (PLC.CkValue  PLC.DefaultUni))
+                          CEK -> PLC.unsafeEvaluateCek (PLC.getStringBuiltinMeanings @ (PLC.CekValue PLC.DefaultUni)) PLC.defaultCostModel
+            body = void . PLC.toTerm $ prog
+            _ = rnf body   -- Force evaluation of body to ensure that we're not timing parsing/deserialisation
```

Comment:
But that doesn't force anything? It should be `!_ = rnf body` at least, otherwise you just introduce a bogus binding that gets ignored. Or am I missing something?

I'd make it `Control.Exception.evaluate $ rnf body` (that should be run in the `IO` monad, not in a `let`).

Reply:
I'm still not sure what's going on there.  I tried `!_ = ...` and various things involving `deepSeq` (both inside and outside the `let`)  when I originally implemented this and nothing seemed to make any difference.  I've just tried it again on an 1100-line PLC program with huge amounts of added spaces to take the overall size to 2.4M.  The program's a big loop that I've set to execute 0 times,  so the main cost should be parsing.  If I leave out the `rnf` stuff altogether it reports 0.05 milliseconds, but if I add `_ <-  Control.Exception.evaluate $ rnf body` then it reports 0.03 or 0.04 ms.  Interestingly the CBOR version if slightly slower: 0.09ms without `rnf`, 0.07 with (and the CBOR file's only 2.2K).  Maybe parsing/deserialisation is just fast, but I'm a bit surprised that the parser's apparently only taking 10 or 20 microseconds to parse a 2.4GB file.

I'll commit the version with `rnf`, but I'm hoping that we can replace the timing code with something from Criterion at some point; at least then we'll be able to blame the Criterion authors for any inaccuracies.

Reply:
[Later]  I put lots of nested comments into my source file, taking the size up to 110MB, and the real-world time for that to execute is about 10 seconds.  If I put `start <- getCPUTime` just before `TypedProgram prog <- getProg Typed inp fmt
` then `plc` reports an execution time of 8931ms.  If I put it just before the line involving `rnf body` then it says 0.22ms, and if I put it just after that then it's 0.11ms.  It looks as if the parser's already returning a fully evaluated AST and `rnf` is probably just doing a traversal where it doesn't have to do any real work. 

For the CBOR version of the same program, it's 0.68 ms if I start recording the time before loading the file, 0.28ms before `rnf`, and 0.07 just before evaluation. 

So it looks as if `rnf` isn't really doing very much anyway.  I feel better now though: what's happening actually makes some sense now.